# 10 Matrix Inverse

<img src="figs/07_winds.png" alt="Drawing" width=400px/>

<img src="figs/10_poly.png" alt="Drawing" width=450px/>

### Why Do We Need Matrix Inverses?

- Because with matrices we do not divide! 
  - Instead, we multiply by the inverse of a matrix, which achieves the same result.

- Because we need it for real-world applications!
  - We use matrix inverses to solve our linear equations.

#### Unit 1: Vectors, Textbook Ch. 1-5

#### Unit 2: Matrices, Textbook Ch. 6-11
- 06 Matrices
- 07 Linear Equations
- 08 Linear Dynamical Systems
- 09 Matrix Multiplication
- **_10 Matrix Inverse_**

#### Unit 3: Least Squares, Textbook Ch. 12-14

##### Outline: 10 Matrix Inverse

- **[Left and right inverses](#sec-matrices)**
- [Inverse](#sec-matrices)
- [Solving linear equations](#sec-matrices)
- [Examples](#sec-matrices)

### Scalar Inverse

$\color{#EF5645}{\text{Definition}}$: Consider a scalar $a$. A scalar $x$ that satisfies $xa = 1$ is called the inverse of $a$.
- We have $x = \frac{1}{a}$, which exists and is unique if and only if $a \neq 0 $.
- We have $xa = ax = 1$.




### Matrix Left Inverse

$\color{#EF5645}{\text{Definition}}$: Consider a matrix $A$. A matrix $X$ that satisfies:
$$XA = I$$ is called a left-inverse of $A$. 

$\color{#EF5645}{\text{Remarks:}}$ 
- **Existence**:  A left inverse might not exist. If a left inverse exists, A is left-invertible. 
- **Uniqueness**: The left-inverse might not be unique.

$\color{#047C91}{\text{Exercise}}$: Show that the matrix:
<center><img src="figs/10_a_to_inverse.png" width=300px alt="default"/></center>

has two different left-inverses:
<center><img src="figs/10_two_left_inverses.png" width=900px alt="default"/></center>


### Properties of left inverses

$\color{#6D7D33}{\text{Properties}}$:
- If $A$ has a left inverse, then the columns of $A$ are linearly independent.
- If $A$ has a left inverse, then $A$ is tall or square.

$\color{#047C91}{\text{Exercise}}$: Prove the above statements.

### Solving linear equations with left inverses

$\color{#EF5645}{\text{Proposition}}$: Consider the linear equation $Ax = b$. Consider $C$ a left-inverse of $A$. Then:
$$x_0 = Cb,$$
is a solution to the linear equation.


$\color{#047C91}{\text{Exercise}}$: Prove the above statement.

$\color{#047C91}{\text{Example}}$: Consider the matrix 
<center><img src="figs/10_a_to_inverse.png" width=300px alt="default"/></center>
from a previous slide, and $b = [1, -2, 0]$. 

Give two solutions to the linear equation:
$$ Ax = b.$$

### Right inverses

$\color{#EF5645}{\text{Definition}}$: Consider a matrix $A$. A matrix $X$ that satistifies:
$$AX = I$$ is called a right-inverse of $A$.

$\color{#EF5645}{\text{Remarks:}}$ 
- **Existence**:  A right inverse might not exist. If a right inverse exists, A is right-invertible. 
- **Uniqueness**: The right-inverse might not be unique.

### Properties of right inverses

$\color{#6D7D33}{\text{Properties}}$:
- $A$ is right invertible if and only if $A^T$ is left invertible.
- $A$ is right invertible if and only if its rows are linearly independent.
- If $A$ is right invertible, then $A$ is wide or square.

$\color{#047C91}{\text{Exercise}}$: Prove the above statements.

### Solving linear equations with right inverses

$\color{#EF5645}{\text{Proposition}}$: Consider the linear equation $Ax = b$. Consider $B$ a right-inverse of $A$. Then:
$$x_0 = Bb,$$
is a solution to the linear equation.

$\color{#047C91}{\text{Exercise}}$: Prove the above statement.

##### Outline: 10 Matrix Inverse

- [Left and right inverses](#sec-matrices)
- **[Inverse](#sec-matrices)**
- [Solving linear equations](#sec-matrices)
- [Examples](#sec-matrices)

### Inverse

$\color{#EF5645}{\text{Definition}}$: If $A$ has a left and a right inverse, then they are equal, and represent the unique inverse of $A$. 

We say that $A$ is invertible. 

We denote $A^{-1}$ the (unique) inverse of $A$.



$\color{#6D7D33}{\text{Properties}}$:
- If $A$ is invertible then $A$ is square.
- The inverse of the inverse is: $(A^{-1})^{-1} = A$.

$\color{#047C91}{\text{Exercise}}$: Prove the above statements.

### Which Matrices are Invertible?

$\color{#6D7D33}{\text{Properties}}$: These matrices are always invertible:
- Any lower triangular matrix $L$ with nonzero diagonal entries is invertible. 
- Any upper triangular $R$ with nonzero diagonal entries is invertible.

$\color{#047C91}{\text{Exercise}}$: Give examples of invertible matrices.

### Computing Inverses: $2 \times 2$ matrices

$\color{#6D7D33}{\text{Properties}}$: Consider $A$ is a $2 \times 2$ matrix:
- $A$ is invertible if and only if $A_{11}A_{22} \neq A_{12}A_{21}$.
- In this case:

<center><img src="figs/lec10_2by2_inverse.png" width=400px alt="default"/></center>


$\color{#047C91}{\text{Exercise}}$: Compute the inverse of:

<center><img src="figs/lec10_exercise_2by2_inverse.png" width=100px alt="default"/></center>


$\color{#047C91}{\text{Answer}}$: https://join.iclicker.com/RR4DX.


A. 

<center><img src="figs/lec10_iclickera.png" width=150px alt="default"/></center>


B.

<center><img src="figs/lec10_iclickerb.png" width=100px alt="default"/></center>

C.

<center><img src="figs/lec10_iclickerc.png" width=100px alt="default"/></center>


###  Computing Inverses

$\color{#6D7D33}{\text{Properties}}$: Some matrices have "easy" inverses:
- The identity matrix: 
  - Then $I^{-1} = I$
- A square $Q$ matrix that verifies $Q^TQ = I$:
  - Then $Q^{-1} = Q^T$.
- A diagonal matrix $D = diag(a_1, ..., a_n)$ with nonzero elements:
  - Then $D^{-1} = diag(\frac{1}{a_1}, ..., \frac{1}{a_n}).$

### Computing Inverses

$\color{#6D7D33}{\text{Properties}}$: Consider invertible square matrices $A, B$ with known inverses $A^{-1}, B^{-1}$.
- $(AB)^{-1} = B^{-1}A^{-1}$
- $(A^T)^{-1} = (A^{-1})^T$

$\color{#047C91}{\text{Exercise}}$: Prove the two first bullet points.


$\color{#EF5645}{\text{Notation}}$: Negative powers! $A^{-k} = (A^{k})^{-1}$

### Computing Inverses: QR decomposition

$\color{#6D7D33}{\text{Properties}}$: Consider $A$, a square and invertible matrix. Consider the QR factorization $A = QR$. 
- Then, the inverse of $A$ can be written: $A^{-1} = R^{-1}Q^T$.

$\color{#EF5645}{\text{Remark}}$: QR is generally not unique. If $A$ is full rank, and we ensure that elements on the diagonal of R are positive, then QR is unique.

### Computing Inverses in Python

$\color{#003660}{\text{In Python}}$, we use `np.linalg.inv` to compute the inverse.

In [7]:
import numpy as np

A = np.array([
    [1, 2],
    [0, 4]
])
print(np.linalg.inv(A))

[[ 1.   -0.5 ]
 [ 0.    0.25]]


##### Outline: 10 Matrix Inverse

- [Left and right inverses](#sec-matrices)
- [Inverse](#sec-matrices)
- **[Solving linear equations](#sec-matrices)**
- [Examples](#sec-matrices)

### Recall: Linear equations

Linear equations translate a real world problem into a math problem.

$\color{#047C91}{\text{Example}}$: An airplane travels 1200 miles in 4 hours with a tail wind. On the way back, the same trip takes 5 hours, now with a head wind (against the wind). To find $v$ the speed of the plane in still air, and $w$ the wind speed, we write the linear equation:

$$1200 = 4 \times (v + w)$$
$$1200 = 5 \times (v - w)$$

This equation is "small" and easy to solve manually. Now, we will see matrix methods to solve more complicated equations!

### Recall: Linear equations

$\color{#EF5645}{\text{Definition}}$: A set (or system) of $m$ linear equations in $n$ variables $x_1, . . . , x_n$ is defined as:
$$\begin{matrix}
A_{11}x_1 + A_{12}x_2 + · · · + A_{1n}x_n = b_1 \\
\vdots \\
A_{m1}x_1 + A_{m2}x_2 + · · · + A_{mn}x_n = b_m
\end{matrix}$$
and can be written compactly as: 
$$Ax = b.$$

### Important Proposition

$\color{#EF5645}{\text{Proposition}}$: If $A$ is invertible with inverse $A^{-1}$, then the equation $Ax = b$ has a unique solution: $x = A^{-1}b$.

$\color{#047C91}{\text{Example}}$: Consider the linear equation from the plane problem:
$$1200 = 4 \times (v + w)$$
$$1200 = 5 \times (v - w)$$
What is $A$? Is it invertible and if so, what is its inverse? Can you solve this linear equation with the above proposition?

In what follows, we see methods to solve $Ax = b$ in several special cases:
- when we can compute $A^{-1}$
- when $A$ is upper-triangular invertible
- when we know the QR decomposition of $A$
- using Python.

### Special case: we can compute $A^{-1}$

$\color{#6D7D33}{\text{Method}}$: Consider $A$ an invertible matrix and the linear equation $Ax = b$. Assume that we know $A^{-1}$. 
- Then the unique solution of $Ax = b$ is given by $A^{-1}b.$

$\color{#047C91}{\text{Example}}$: The plane exercise we just did.

### Special case: $A = R$ upper triangular invertible

$\color{#6D7D33}{\text{Method}}$:  Consider $R$ an upper triangular matrix with nonzero entries and the linear equation: $Rx = b$, which can be re-written as:
$$\begin{matrix}
R_{11}x_1 + & R_{12}x_2 + & ... +& R_{1, n}x_n &= b_1 \\
&&\vdots&&\\
&&& R_{nn}x_n &= b_n
\end{matrix}$$

The solution of the linear equation can be found by back-substitution:
- Last equation gives: $x_n = b_n / R_{nn}$
- Second to last equation gives: $x_{n-1} = (b_{n-1} - R_{n-1, n}x_n)/ R_{n-1, n-1}$
- Iterate.

### Special case: $A = QR$ via QR Factorization

$\color{#6D7D33}{\text{Method}}$: Consider $A$ an invertible matrix and the linear equation $Ax = b$. Assume that the $QR$ factorization of $A$ is given: $A = QR$.

The solution of the linear equation can be found by using these steps:
- Compute $Q^T b$
- Solve the linear equation $Rx = Q^Tb$ by back substitution.


### General case in Python

$\color{#003660}{\text{In Python}}$, a linear equation can be solved in several ways:
- using `np.linalg.solve`
- using `np.linalg.inv`

In [3]:
A = np.array([[-3, -4], [4, 6]]); b = np.array([1, 2])
print(np.linalg.solve(A, b)); print(np.linalg.inv(A) @ b)

[-7.  5.]
[-7.  5.]


##### Outline: 10 Matrix Inverse


- [Left and right inverses](#sec-matrices)
- [Inverse](#sec-matrices)
- [Solving linear equations](#sec-matrices)
- **[Examples](#sec-matrices)**

### Example: Interpolation

In [1]:
from IPython.display import Video; Video("figs/11_aftereff.mp4")

<center><img src="figs/11_interpolation.png" alt="default"/></center>

### Polynomial Interpolation

$\color{#047C91}{\text{Exercise}}$: Consider a cubic polynomial with unknown coefficients $c_0, ..., c_3$:
$$p(x) = c_0 + c_1 x + c_2 x^2 + c_3 x^3,$$
that satisfies: $p(-1.1) = 1, p(-0.4) = 2, p(0.1)=4, p(0.8) = 1$.

Find the polynomial interpolating these 4 points using Python, i.e. find the equation of the blue curve below.

<center><img src="figs/10_poly.png" alt="default" width=300px/></center>

In [5]:
A = np.array([
    [1, -1.1, (-1.1) ** 2, (-1.1) ** 3],
    [1, -0.4, (-0.4) ** 2, (-0.4) ** 3],
    [1, 0.1, (0.1) ** 2, (0.1) ** 3],
    [1, 0.8, (0.8) ** 2, (0.8) ** 3]
])
b = np.array([1, 2, 4, 1])

np.linalg.solve(A, b)

array([ 3.72380952,  3.26190476, -4.52380952, -4.76190476])

The polynomial is:
$$p(x) = 3.72 + 3.26 x + -4.52 x^2 + -4.76 x^3,$$

##### Outline: 10 Matrix Inverse

- [Left and right inverses](#sec-matrices)
- [Inverse](#sec-matrices)
- [Solving linear equations](#sec-matrices)
- [Examples](#sec-matrices)

Resources: Textbook Ch. 11